src/consensus/event_accumulator.rs

```diff
@@ -80,194 +90,168 @@ impl VoteStatuses {
-                    event.into_network_event_with(proofs.into_sig_shares().remove(our_id))
-                })
+                .filter(|(_, state)| state.voters.contains(our_name))
+                .map(|((event, _), _)| event)
                 .collect(),
             accumulated_events,
         }
     }
 
-    pub fn add_expectation(
-        &mut self,
-        event: AccumulatingEvent,
-        proofs: &AccumulatingProof,
-        all_voters: BTreeSet<PublicId>,
-    ) {
-        let mut non_voted = all_voters.clone();
-        for id in proofs.parsec_proof_set().ids() {
-            let _ = non_voted.remove(id);
-        }
-        self.vote_statuses
-            .add_expectation(event, non_voted, &all_voters);
-    }
-
-    pub fn detect_unresponsive<'a>(
-        &self,
-        members: impl IntoIterator<Item = &'a PublicId>,
-    ) -> BTreeSet<PublicId> {
-        members
-            .into_iter()
-            .filter(|peer_id| self.vote_statuses.is_unresponsive(peer_id))
-            .cloned()
+    pub fn detect_unresponsive(&self, elders_info: &EldersInfo) -> BTreeSet<PublicId> {
+        elders_info
+            .elder_ids()
+            .filter(|id| self.vote_statuses.is_unresponsive(id.name()))
+            .copied()
             .collect()
     }
 }
 
+// Accumulation state of a single event in the event accumulator.
 #[derive(Default, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Serialize, Deserialize)]
-pub struct AccumulatingProof {
-    parsec_proofs: ProofSet,
-    sig_shares: BTreeMap<PublicId, bls::SignatureShare>,
+struct State {
+    voters: BTreeSet<XorName>,
+    shares: BTreeMap<usize, bls::SignatureShare>,
 }
 
-impl AccumulatingProof {
-    #[cfg(test)]
-    pub fn from_proof_set(parsec_proofs: ProofSet) -> Self {
-        Self {
-            parsec_proofs,
-            sig_shares: Default::default(),
-        }
-    }
-
-    /// Return false if share or proof is replaced
-    pub fn add_proof(&mut self, proof: Proof, sig_share: Option<bls::SignatureShare>) -> bool {
-        let new_share = sig_share.map_or(true, |share| {
-            self.sig_shares.insert(proof.pub_id, share).is_none()
-        });
+impl State {
+    fn new(voter_name: XorName, voter_index: usize, signature_share: bls::SignatureShare) -> Self {
+        let voters = iter::once(voter_name).collect();
+        let shares = iter::once((voter_index, signature_share)).collect();
 
-        let new_proof = self.parsec_proofs.add_proof(proof);
-        new_share && new_proof
+        Self { voters, shares }
     }
 
-    pub fn parsec_proof_set(&self) -> &ProofSet {
-        &self.parsec_proofs
-    }
+    fn add(
+        &mut self,
+        voter_name: XorName,
+        voter_index: usize,
+        signature_share: bls::SignatureShare,
+    ) -> Result<(), AccumulatingError> {
+        if self.shares.insert(voter_index, signature_share).is_some() {
+            return Err(AccumulatingError::ReplacedAlreadyInserted);
```

Comment:
Yeah, I'm not sure this error even makes sense. Because the only way a vote could be overwritten is with same vote. Otherwise the verification would fail. And if it's the same vote, there is no harm overwriting it. Will try to remove it.